File: example-app-java/src/main/java/io/adtrace/examples/InAppLogger.java
package io.adtrace.examples;

import io.adtrace.sdk.ILogger;
import io.adtrace.sdk.LogLevel;
import io.adtrace.sdk.Logger;
import io.adtrace.sdk.Util;

/**
 * Forwards AdTrace SDK logs to Logcat and mirrors them in {@link AdTraceLogStore}
 * so developers can inspect sent requests inside the example app.
 *
 * Install before creating {@link io.adtrace.sdk.AdTraceConfig}:
 * <pre>
 * AdTraceFactory.setLogger(new InAppLogger());
 * </pre>
 */
public class InAppLogger implements ILogger {

    private final Logger delegate = new Logger();

    @Override
    public void setLogLevel(LogLevel logLevel, boolean isProductionEnvironment) {
        delegate.setLogLevel(logLevel, isProductionEnvironment);
    }

    @Override
    public void setLogLevelString(String logLevelString, boolean isProductionEnvironment) {
        delegate.setLogLevelString(logLevelString, isProductionEnvironment);
    }

    @Override
    public void verbose(String message, Object... parameters) {
        delegate.verbose(message, parameters);
        mirror("V", message, parameters);
    }

    @Override
    public void debug(String message, Object... parameters) {
        delegate.debug(message, parameters);
        mirror("D", message, parameters);
    }

    @Override
    public void info(String message, Object... parameters) {
        delegate.info(message, parameters);
        mirror("I", message, parameters);
    }

    @Override
    public void warn(String message, Object... parameters) {
        delegate.warn(message, parameters);
        mirror("W", message, parameters);
    }

    @Override
    public void warnInProduction(String message, Object... parameters) {
        delegate.warnInProduction(message, parameters);
        mirror("W", message, parameters);
    }

    @Override
    public void error(String message, Object... parameters) {
        delegate.error(message, parameters);
        mirror("E", message, parameters);
    }

    @Override
    public void Assert(String message, Object... parameters) {
        delegate.Assert(message, parameters);
        mirror("A", message, parameters);
    }

    @Override
    public void lockLogLevel() {
        delegate.lockLogLevel();
    }

    private void mirror(String level, String message, Object... parameters) {
        if (message == null) {
            return;
        }
        String formatted;
        try {
            formatted = Util.formatString(message, parameters);
        } catch (Exception e) {
            formatted = message;
        }
        AdTraceLogStore.append("[" + level + "] " + formatted);
    }
}

File: example-app-kotlin/src/main/java/io/adtrace/examples/InAppLogger.kt
package io.adtrace.examples

import io.adtrace.sdk.ILogger
import io.adtrace.sdk.LogLevel
import io.adtrace.sdk.Logger
import io.adtrace.sdk.Util

/**
 * Forwards AdTrace SDK logs to Logcat and mirrors them in [AdTraceLogStore]
 * so developers can inspect sent requests inside the example app.
 *
 * Install before creating [io.adtrace.sdk.AdTraceConfig]:
 * ```
 * AdTraceFactory.setLogger(InAppLogger())
 * ```
 */
class InAppLogger : ILogger {

    private val delegate = Logger()

    override fun setLogLevel(logLevel: LogLevel, isProductionEnvironment: Boolean) {
        delegate.setLogLevel(logLevel, isProductionEnvironment)
    }

    override fun setLogLevelString(logLevelString: String?, isProductionEnvironment: Boolean) {
        delegate.setLogLevelString(logLevelString, isProductionEnvironment)
    }

    override fun verbose(message: String?, vararg parameters: Any?) {
        delegate.verbose(message, *parameters)
        mirror("V", message, parameters)
    }

    override fun debug(message: String?, vararg parameters: Any?) {
        delegate.debug(message, *parameters)
        mirror("D", message, parameters)
    }

    override fun info(message: String?, vararg parameters: Any?) {
        delegate.info(message, *parameters)
        mirror("I", message, parameters)
    }

    override fun warn(message: String?, vararg parameters: Any?) {
        delegate.warn(message, *parameters)
        mirror("W", message, parameters)
    }

    override fun warnInProduction(message: String?, vararg parameters: Any?) {
        delegate.warnInProduction(message, *parameters)
        mirror("W", message, parameters)
    }

    override fun error(message: String?, vararg parameters: Any?) {
        delegate.error(message, *parameters)
        mirror("E", message, parameters)
    }

    override fun Assert(message: String?, vararg parameters: Any?) {
        delegate.Assert(message, *parameters)
        mirror("A", message, parameters)
    }

    override fun lockLogLevel() {
        delegate.lockLogLevel()
    }

    private fun mirror(level: String, message: String?, parameters: Array<out Any?>) {
        if (message == null) {
            return
        }
        val formatted = try {
            Util.formatString(message, *parameters)
        } catch (e: Exception) {
            message
        }
        AdTraceLogStore.append("[$level] $formatted")
    }
}
